File: count-of-smaller-numbers-after-self.cpp
#define N 100005
class Solution {
public:
    int a[N];
    vector<int> tmp;
    vector<int> res;
    int n;
    int findIndex(int val) {
        int left = 0, right = n-1;
        while (left < right) {
            int mid = (left + right) / 2;
            if (tmp[left] == val || tmp[right] == val || tmp[mid] == val) {
                if (tmp[left] == val) return left + 1;
                if (tmp[right] == val) return right + 1;
                if (tmp[mid] == val) return mid + 1;
            }
            if (val < tmp[mid]) right = mid;
            else left = mid;
        }
        return left + 1;
    }
    void addValue(int begin) {
        if (begin <= 0) {
            while(true);
            return;
        }
        for (int idx = begin; idx < N; idx += idx&(-idx)) {
            a[idx] += 1;
        }
    }
    int getValue(int index) {
        int sum = 0;
        for (int idx = index; idx > 0; idx -= idx&(-idx)) {
            sum += a[idx];
        }
        return sum;
    }
    vector<int> countSmaller(vector<int>& nums) {
        n = nums.size();
        tmp.clear();
        for (int i = 0; i < nums.size(); ++i) {
            tmp.push_back(nums[i]);
            res.push_back(0);
            a[i+1] = 0;
        }
        sort(tmp.begin(), tmp.end());
        for (int i = nums.size() - 1; i >= 0; --i) {
            int idx = findIndex(nums[i]);
            addValue(idx);
            res[i] = getValue(idx - 1);
        }
        return res;
    }
};

File: count-of-smaller-numbers-after-self.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class Solution {
    public Solution() {
    }
    private List<Integer> a = new ArrayList<Integer>();
    private List<Integer> tmp = new ArrayList<Integer>();
    private int n = 0;
    public List<Integer> countSmaller(int[] nums) {
        a.clear();
        tmp.clear();
        n = nums.length;
        List<Integer> res = new ArrayList<Integer>();
        a.add(0);
        for (int val : nums) {
            a.add(0);
            tmp.add(val);
            res.add(0);
        }
        Collections.sort(tmp);
        for (int i = nums.length -1; i >= 0; --i) {
            int idx = getIndex(nums[i]);
            addValue(idx);
            res.set(i, getValue(idx - 1));
        }
        return res;
    }

    private void addValue(int idx) {
        for (int i = idx; i < a.size(); i += i&(-i)) {
            a.set(i, a.get(i) + 1);
        }
    }
    private int getValue(int idx) {
        int sum = 0;
        if (idx <= 0) return 0;
        for (int i = idx; i > 0; i -= i&(-i)) {
            sum += a.get(i);
        }
        return sum;
    }

    private int getIndex(int val) {
        int left = 0, right = n-1, times = 0;
        while (left < right && times <= 65) {
            ++times;
            int mid = (left + right) / 2;
            if (tmp.get(left) == val || tmp.get(right) == val || tmp.get(mid) == val) {
                if (tmp.get(left) == val) return left + 1;
                if (tmp.get(right) == val) return right + 1;
                if (tmp.get(mid) == val) return mid + 1;
            }
            if (val < tmp.get(mid)) right = mid;
            else left = mid;
        }
        return left + 1;
    }

    public void main() {
        int[] nums = new int[] { 9, 6, 1, 8 };
        System.out.println(new Solution().countSmaller(nums));
    }

    public static void main(String[] args) {
        new Solution().main();
    }
}